// entity-finder/lib/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var debug = require('debug')('entity-finder');
var wiki_entity_1 = require("wiki-entity");
var find_titles_1 = require("./find_titles");
var filters_1 = require("./filters");
function find(name, lang, options) {
    if (options === void 0) { options = {}; }
    var limit = options.limit || 2;
    return find_titles_1.findTitles(name, lang, { limit: limit + 2, tags: options.tags })
        .then(function (foundTitles) {
        if (foundTitles.length === 0) {
            return [];
        }
        var titles = foundTitles.map(function (title) { return title.title; });
        debug('finding titles', titles);
        return wiki_entity_1.getEntities({
            titles: titles.join('|'),
            language: lang,
            claims: options.claims,
            extract: options.extract,
            redirects: options.redirects,
            types: options.types
        })
            .then(filters_1.filterWikiEntities)
            .then(function (entities) { return entities.slice(0, limit); });
    });
}
exports.find = find;
