// entity-finder/lib/wikipedia/api.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
var request_1 = require("../request");
var OPTIONS = {
    qs: {
        format: 'json'
    }
};
function createOptions(lang, qs) {
    var options = {
        qs: utils_1._.defaults({}, qs || {}, OPTIONS.qs),
        url: 'https://' + lang + '.wikipedia.org/w/api.php'
    };
    return options;
}
function query(lang, qs) {
    qs.action = 'query';
    var options = createOptions(lang, qs);
    return request_1.default(options);
}
exports.query = query;
;
function openSearch(lang, search, opts) {
    if (opts === void 0) { opts = {}; }
    var qs = {
        search: search,
        action: 'opensearch',
        redirects: opts.redirects || 'resolve',
        suggest: true,
        profile: opts.profile || 'normal',
        limit: opts.limit || 10
    };
    var options = createOptions(lang, qs);
    return request_1.default(options);
}
exports.openSearch = openSearch;
;
function search(lang, srsearch) {
    var qs = {
        srsearch: srsearch,
        list: 'search',
        srprop: 'size'
    };
    return query(lang, qs);
}
exports.search = search;
;
